haskell/private/actions/link.bzl

```diff
@@ -82,42 +87,60 @@ def create_rpath_entry(binary, dependency, keep_filename, prefix = ""):
     Args:
       binary: target of current binary
       dependency: target of dependency to relatively point to
-      prefix: string path prefix to add before the relative path
       keep_filename: whether to point to the filename or its parent dir
+      comes_from_haskell_cc_import: if dependency is a haskell_cc_import
+      prefix: string path prefix to add before the relative path
 
     Returns:
       relative path string
     """
-    (bin_is_external, bin_parent_dir) = _get_target_parent_dir(binary)
-    (dep_is_external, dep_parent_dir) = _get_target_parent_dir(dependency)
-
-    # backup through parent directories of the binary
-    bin_backup = [".."] * len(bin_parent_dir)
-
-    # external repositories live in `target.runfiles/external`,
-    # while the internal repository lives in `target.runfiles`.
-    # The `.short_path`s of external repositories are strange,
-    # they start with `../`, but you cannot just append that in
-    # order to find the correct runpath. Instead you have to use
-    # the following logic to construct the correct runpaths:
-    if bin_is_external:
-        if dep_is_external:
-            # stay in `external`
-            path_segments = bin_backup
-        else:
-            # backup out of `external`
-            path_segments = [".."] + bin_backup
-    elif dep_is_external:
-        # go into `external`
-        path_segments = bin_backup + ["external"]
+
+    # we need to use different paths if the dependency was imported
+    # via the `haskell_cc_import` rule.
+    # XXX: remove once `haskell_cc_import` is removed.
+    if comes_from_haskell_cc_import:
+        # backup the full path to the binary
+        # This abuses the fact that the linker resolves symlinks,
+        # meaning the `binary` path is not actually the path in the
+        # runfiles folder, rather the one in the bazel execpath.
+        bin_backup = [".."] * len(parent_dir_path(binary.path))
+
+        # append the full path of our dependency
+        path_segments = bin_backup + parent_dir_path(dependency.path)
+
+        # external repositories live in `target.runfiles/external`,
```

Comment:
Is this comment in the right place?

Reply:
Good catch, must have been moved by buildifier.